clash-prelude/src/Clash/Explicit/RAM.hs

```diff
@@ -132,32 +135,55 @@ asyncRam#
   -- ^ Value to write (at address @w@)
   -> Signal rdom a
   -- ^ Value of the @RAM@ at address @r@
-asyncRam# wclk rclk en sz rd we wr din =
-    unsafeSynchronizer wclk rclk dout
+asyncRam# !_ !_ en sz rd we wr din = dout
   where
-    rd'  = unsafeSynchronizer rclk wclk rd
     ramI = Seq.replicate
-              (snatToNum sz)
+              szI
               (withFrozenCallStack (errorX "asyncRam#: initial value undefined"))
-    en' = fromEnable (andEnable en we)
-    dout = go ramI rd' en' wr din
+    en0 = fromEnable (andEnable en we)
+    dout = if rPeriod == wPeriod
+           then goSingle ramI rd en0 wr din
+           else go 0 ramI rd en0 wr din
+    rPeriod = snatToNum (clockPeriod @rdom) :: Int
+    wPeriod = snatToNum (clockPeriod @wdom) :: Int
     szI = snatToNum sz :: Int
 
-    go :: Seq.Seq a -> Signal wdom Int -> Signal wdom Bool
-       -> Signal wdom Int -> Signal wdom a -> Signal wdom a
-    go !ram (r :- rs) (e :- es) (w :- ws) (d :- ds) =
-      let ram' = upd ram e (fromEnum w) d
+    goSingle :: Seq.Seq a -> Signal rdom Int -> Signal wdom Bool
+       -> Signal wdom Int -> Signal wdom a -> Signal rdom a
+    goSingle !ram (r :- rs) ~(e :- es) ~(w :- ws) ~(d :- ds) =
+      let ram0 = upd ram e w d
           o    = ram `safeAt` r
-      in  o :- go ram' rs es ws ds
-
-    upd ram we' waddr d = case maybeIsX we' of
-      Nothing -> case maybeIsX waddr of
-        Nothing -> fmap (const (seq waddr d)) ram
-        Just wa -> safeUpdate wa d ram
-      Just True -> case maybeIsX waddr of
-        Nothing -> fmap (const (seq waddr d)) ram
-        Just wa -> safeUpdate wa d ram
-      _ -> ram
+      in  o :- goSingle ram0 rs es ws ds
+
+    -- Given
+    --   tR = absolute time of next active edge of read clock
+    --   tW = absolute time of next active edge of write clock
+    -- relTime is defined as relTime = tW - tR
+    --
+    -- Put differently, relative time 0 points at the next active edge of the
+    -- read clock, and relTime points at the next active edge of the write
+    -- clock.
+    go :: Int -> Seq.Seq a -> Signal rdom Int -> Signal wdom Bool
+       -> Signal wdom Int -> Signal wdom a -> Signal rdom a
+    go   relTime !ram rt@(~(r :- rs)) et@(~(e :- es)) wt@(~(w :- ws))
+         dt@(~(d :- ds))
+      | relTime < 0 = let ram0 = upd ram e w d
+                      in go (relTime + wPeriod) ram0 rt es ws ds
+      | otherwise   = let o = ram `safeAt` r
+                      in o :- go (relTime - rPeriod) ram rs et wt dt
+
+    upd ram e waddr d = case maybeIsX waddr of
+      Nothing -> case maybeIsX e of
+                   Just False -> ram
+                   _ ->  -- Put the XException from `waddr` as the value in all
+                         -- locations of `ram`.
+                         fmap (const (seq waddr d)) ram
+      Just wa -> case maybeIsX e of
+        Nothing -> -- Put the XException from `e` as the value at address
```

Comment:
I'd like to reorder this again, because we don't need to evaluate `waddr` if `e` is `False`. I ordered it to avoid duplicating code, but this code duplication provides a speed benefit if people have an expensive `waddr`.